Add ref for relu6, fixes hardshrink and improves testing of related ops

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #81142

This PR:
- Adds ref for relu6 and makes its OpInfo a UnaryUfuncInfo
- Correct hardshrink ref when lambd < 0 and when inputs are nan
- Corrected nan behavior vectorized implementation of hardshrink (fixes https://github.com/pytorch/pytorch/issues/81138)
- Make OpInfos for {hard,soft}shrink, hardtanh UnaryUfuncInfos and add error_inputs for softshrink

soulitzer committed:
Add ref for relu and improve related ops and testing

[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for relu and improve related ops and testing"




[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for relu and improve related ops and testing"


This PR:
- Adds ref for relu6
- Make {hard,soft}shrink, hardtanh UnaryUfuncs and add error_inputs for softshrink
- Corrected nan behavior vectorized implementation of harshrink
- Correct hardshrink ref when lambd < 0 

[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for relu6, fixes hardshrink and improves testing of related ops"


This PR:
- Adds ref for relu6 and makes its OpInfo a UnaryUfuncInfo
- Correct hardshrink ref when lambd < 0 and when inputs are nan
- Corrected nan behavior vectorized implementation of hardshrink (fixes https://github.com/pytorch/pytorch/issues/81138)
- Make OpInfos for {hard,soft}shrink, hardtanh UnaryUfuncInfos and add error_inputs for softshrink


[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for relu6, fixes hardshrink and improves testing of related ops"


This PR:
- Adds ref for relu6 and makes its OpInfo a UnaryUfuncInfo
- Correct hardshrink ref when lambd < 0 and when inputs are nan
- Corrected nan behavior vectorized implementation of hardshrink (fixes https://github.com/pytorch/pytorch/issues/81138)
- Make OpInfos for {hard,soft}shrink, hardtanh UnaryUfuncInfos and add error_inputs for softshrink


[ghstack-poisoned]

lezcano reviewed (approved):
Cool. It's true that we shouldn't use branchless programming unless it's a part of the code that's crucial, as otherwise we end up hitting this edge cases over and over.

soulitzer committed:
Update on "Add ref for relu6, fixes hardshrink and improves testing of related ops"


This PR:
- Adds ref for relu6 and makes its OpInfo a UnaryUfuncInfo
- Correct hardshrink ref when lambd < 0 and when inputs are nan
- Corrected nan behavior vectorized implementation of hardshrink (fixes https://github.com/pytorch/pytorch/issues/81138)
- Make OpInfos for {hard,soft}shrink, hardtanh UnaryUfuncInfos and add error_inputs for softshrink


[ghstack-poisoned]

Lezcano commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -270,7 +270,7 @@ def hardshrink(a: TensorLikeType, lambd: float = 0.5):
     #               = x if x < -lambd
     #               = 0 otherwise
-    return refs.where(abs(a) > abs(lambd), a, 0)
+    return refs.where(refs.logical_and(a >= -lambd, a <= lambd), 0, a)
```
I think there's no NaN test for UnaryUfuncs, so ref inputs would be good.

soulitzer replied:
Yup, UnaryUfuncInfo automatically adds `reference_inputs_elementwise_unary` which adds the NaN testing

Lezcano replied:
ooops. Thanks for the correction!

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -485,3 +489,13 @@ def prelu(a: TensorLikeType, weight: TensorLikeType) -> TensorLikeType:
+
+def relu6(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
+    """
+    Reference implementation of torch.nn.functional.prelu
```
of relu6, not prelu?

soulitzer replied:
good catch, whoops!

soulitzer committed:
Update on "Add ref for relu6, fixes hardshrink and improves testing of related ops"


This PR:
- Adds ref for relu6 and makes its OpInfo a UnaryUfuncInfo
- Correct hardshrink ref when lambd < 0 and when inputs are nan
- Corrected nan behavior vectorized implementation of hardshrink (fixes https://github.com/pytorch/pytorch/issues/81138)
- Make OpInfos for {hard,soft}shrink, hardtanh UnaryUfuncInfos and add error_inputs for softshrink


[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -7297,12 +7297,34 @@ def sample_inputs_linalg_svdvals(op_info, device, dtype, requires_grad=False, **
     for batch, m, n in product(batches, ns, ns):
         yield SampleInput(make_arg(batch + (m, n)))
 
+def error_inputs_softshrink(op, device, **kwargs):
```
Nice error input

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -7297,12 +7297,34 @@ def sample_inputs_linalg_svdvals(op_info, device, dtype, requires_grad=False, **
+def sample_inputs_softshrink(op_info, device, dtype, requires_grad=False, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    for lbda in (0., 0.5):
```
Add a brief comment for each of these about why the addition sample is added to the elementwise samples

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -20451,7 +20468,7 @@ def __init__(
         torch_opinfo_name="nn.functional.elu",
     ),
-    PythonRefInfo(
+    ElementwiseUnaryPythonRefInfo(
```
Awesome updates to extend the testing

mruberry reviewed (approved):
This is awesome, @soulitzer -- I made a few comments inline for your review

soulitzer commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -485,3 +489,13 @@ def prelu(a: TensorLikeType, weight: TensorLikeType) -> TensorLikeType:
+    if inplace:
+        raise NotImplementedError
+
+    return refs.nn.functional.hardtanh(a, 0, 6)
```
Oh interesting, I just used hardtanh here because that was the implementation does. We would want to be consistent with that right?

mruberry replied:
I'd suggest using `clamp` because primTorch is an opportunity to improve the quality of implementations, but some reference implementers prefer fidelity since fidelity is often faster to implement, so there's no hard rule.

Either way I'd add a brief comment here about which choice you made and why

soulitzer committed:
Update on "Add ref for relu6, fixes hardshrink and improves testing of related ops"


This PR:
- Adds ref for relu6 and makes its OpInfo a UnaryUfuncInfo
- Correct hardshrink ref when lambd < 0 and when inputs are nan
- Corrected nan behavior vectorized implementation of hardshrink (fixes https://github.com/pytorch/pytorch/issues/81138)
- Make OpInfos for {hard,soft}shrink, hardtanh UnaryUfuncInfos and add error_inputs for softshrink


[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for relu6, fixes hardshrink and improves testing of related ops"


This PR:
- Adds ref for relu6 and makes its OpInfo a UnaryUfuncInfo
- Correct hardshrink ref when lambd < 0 and when inputs are nan
- Corrected nan behavior vectorized implementation of hardshrink (fixes https://github.com/pytorch/pytorch/issues/81138)
- Make OpInfos for {hard,soft}shrink, hardtanh UnaryUfuncInfos and add error_inputs for softshrink


[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g
